flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java

```diff
@@ -107,43 +109,42 @@ public void testErrorReporting() throws Exception {
 
     @Test
     public void checkScheduledTimestamps() throws Exception {
-        final AtomicReference<Throwable> errorRef = new AtomicReference<>();
-
-        final long t1 = System.currentTimeMillis();
-        final long t2 = System.currentTimeMillis() - 200;
-        final long t3 = System.currentTimeMillis() + 100;
-        final long t4 = System.currentTimeMillis() + 200;
-
-        timeService.registerTimer(t1, new ValidatingProcessingTimeCallback(errorRef, t1, 0));
-        timeService.registerTimer(t2, new ValidatingProcessingTimeCallback(errorRef, t2, 1));
-        timeService.registerTimer(t3, new ValidatingProcessingTimeCallback(errorRef, t3, 2));
-        timeService.registerTimer(t4, new ValidatingProcessingTimeCallback(errorRef, t4, 3));
-
-        long deadline = System.currentTimeMillis() + 20000;
-        while (errorRef.get() == null
-                && ValidatingProcessingTimeCallback.numInSequence < 4
-                && System.currentTimeMillis() < deadline) {
-            Thread.sleep(100);
+        ValidatingProcessingTimeCallback.numInSequence = 0;
+        long currentTimeMillis = System.currentTimeMillis();
+        ArrayList<ValidatingProcessingTimeCallback> timeCallbacks = new ArrayList<>();
+
+        /*
+         It is not possible to register timer for currentTimeMillis or value slightly greater than
+         currentTimeMillis because if the during registerTimer the internal currentTime is equal
+         to this value then according to current logic the time will be increased for 1ms while
+         `currentTimeMillis - 200` is always transform to 0, so it can lead to reordering. See
+         comment in {@link ProcessingTimeServiceUtil#getRecordProcessingTimeDelay(long, long)}.
+        */
+        timeCallbacks.add(new ValidatingProcessingTimeCallback(currentTimeMillis - 1, 0));
+        timeCallbacks.add(new ValidatingProcessingTimeCallback(currentTimeMillis - 200, 1));
+        timeCallbacks.add(new ValidatingProcessingTimeCallback(currentTimeMillis + 100, 2));
+        timeCallbacks.add(new ValidatingProcessingTimeCallback(currentTimeMillis + 200, 3));
+
+        for (ValidatingProcessingTimeCallback timeCallback : timeCallbacks) {
+            timeService.registerTimer(timeCallback.expectedTimestamp, timeCallback);
         }
 
-        verifyNoException(errorRef.get());
+        for (ValidatingProcessingTimeCallback timeCallback : timeCallbacks) {
+            timeCallback.assertExpectedValues();
+        }
         assertEquals(4, ValidatingProcessingTimeCallback.numInSequence);
     }
 
     private static class ValidatingProcessingTimeCallback implements ProcessingTimeCallback {
 
         static int numInSequence;
 
-        private final AtomicReference<Throwable> errorRef;
+        private final CompletableFuture<Void> finished = new CompletableFuture<>();
```

Comment:
The bug fix in the test and the refactoring of this test I think are independent and should be in separate commits?